### ISSUE TITLE
sql: intentionally randomize the order of results when no ORDER BY is specified

### ISSUE THREAD

**Issue by rjnn** [cockroachdb/cockroach#24223]
Currently, CockroachDB queries that intuitively would scan in some natural order don't guarantee to hold to that order. This is consistent with the SQL standard, which does not require that queries return results in order unless an `ORDER BY` is specified. Most single node databases _do_ return results in order, as does cockroach when the results are on a single node. This is confounded by the fact that the `EXPLAIN` plans for a query with an `ORDER BY` looks identical to a query that happens to return in that order, so developers often look at the plan to ensure it's optimized, and then neglect to add a final ORDER BY, since the plans are "identical".

Consider an example, selecting by a table's primary key column:

```
root@:26257/tpcc> explain select * from district order by d_w_id limit 10;
+-----------+-------+------------------+
|   Tree    | Field |   Description    |
+-----------+-------+------------------+
| limit     |       |                  |
|  └── scan |       |                  |
|           | table | district@primary |
|           | spans | ALL              |
|           | limit | 10               |
+-----------+-------+------------------+
(5 rows)

root@:26257/tpcc> explain select * from district limit 10;
+-----------+-------+------------------+
|   Tree    | Field |   Description    |
+-----------+-------+------------------+
| limit     |       |                  |
|  └── scan |       |                  |
|           | table | district@primary |
|           | spans | ALL              |
|           | limit | 10               |
+-----------+-------+------------------+
(5 rows)
```

The real problem is that without an explicit `ORDER BY`, when the table is distributed, cockroachdb optimizes results delivery by returning the first values it receives, regardless of order. Thus, users often encounter this artifact long after the fact, as it "always works" in testing on small data sizes, and only fails at scale.

One solution is to throw in a little bit of randomization when no `ORDER BY` is specified. Care should be taken to not do so at the cost of performance (perhaps only do this for a small fraction of rows). [This is what golang does](https://nathanleclaire.com/blog/2014/04/27/a-surprising-feature-of-golang-that-colored-me-impressed/), and it helps build the intuition early that results won't be ordered unless you explictly specify that they are.

I would like to do the same in CockroachDB, unless I get strong pushback that this is an anti-feature. One useful data point is that this wrong ordering assumption has already bitten users (including some engineers who work on CockroachDB such as myself and @mjibson).

**Comment by jordanlewis** [comment]
I still don't think we should enable this by default. While something like this might make intuitive sense to experienced programmers, we should not only be targeting experienced programmers as potential users of the database. We should be kind to our users. I understand that a feature like this is meant to be kind to users in the long run, because it trains them that results are unpredictable without an `ORDER BY` statement. However, in my opinion, this kind of training is user-hostile. It reminds me of passive-aggressive behavior - instead of directly telling the user that they should not expect ordered results without `ORDER BY`, this proposal hands them artificially randomized results and expects them to figure out what is going on themselves.

In short, I think it "makes data harder".

----

Other people have suggested hiding this behind a session flag and enabling it always in tests. I think both of those are great ideas.

I would further suggest that we fix the confusing `EXPLAIN` diagram that you mention. You argue that the lack of change between the two diagrams confounds understandability - I agree, but the fix shouldn't have to be introducing artificial randomization of database results. Instead, we should fix the diagram: it should include an `ORDER BY` that's marked as implicit, or something like that.

**Comment by arjunravinarayan** [comment]
You're right that augmenting the EXPLAIN plans, and hiding this behind a flag is probably a better way to go.

**Comment by bdarnell** [comment]
I think if this is off by default it will be so rarely enabled that it's not worth doing - anyone who is aware enough of this issue to enable the setting is already way better off than most users. If we do this it should be on by default (or even mandatory - I see no reason to make it configurable).

**Comment by rmloveland** [comment]
If I understand what I read above correctly, this change will make the user experience at the SQL prompt less nice than it currently is.

It protects users from relying on unspecified behavior but imposes a cost on the new or less advanced user, who will henceforth have to type out the `ORDER BY` in many queries.  If the user base grows as large as we hope, this may result in a lot of keystrokes being spent over the years.  I think we should weigh the cost of that user effort against the potential benefits of this proposed change.

As mentioned we already [document](https://www.cockroachlabs.com/docs/dev/select.html) the lack of ordering:

> When ORDER BY is not included in a query, rows are not sorted by any consistent criteria. Instead, CockroachDB returns them as the coordinating node receives them.

Our current contract as specified by our docs may make a change like this "legal", but as a user I would not want this change.  There is a vast psychological gulf (IMO) between "it works this way, but don't rely on it, especially in these cases" and "we intentionally changed it, we're allowed to do that by the spec".  I think the latter feels not very nice to be on the other side of without a strong justification, even if that's admittedly not entirely rational.

As an example from another context, as a user I am willing to endure some of the strange behaviors performed by C compilers that are still "within spec" if they get me something I want (in that case speed, theoretically?), whereas this doesn't seem like it is buying me something I want as a user, it seems to be enforcing a particular usage pattern.

Finally, if this sort of intentional randomization is not the default behavior of other database products we will have one more difference to explain, one more hurdle to adoption.  Possibly a tiny hurdle but one we would be putting in front of the potential user or customer intentionally.

**Comment by bdarnell** [comment]
> It protects users from relying on unspecified behavior but imposes a cost on the new or less advanced user, who will henceforth have to type out the ORDER BY in many queries.

You don't have to type out the ORDER BY *unless you rely on the order*. If you're just messing around at the prompt and don't care about the order, you can continue typing your queries without ORDER BY.

**Comment by rmloveland** [comment]
> If you're just messing around at the prompt and don't care about the order, you can continue typing your queries without ORDER BY

**Comment by mjibson** [comment]
I am strongly for this proposal. I'll try to address the concerns as I understand them.

Jordan said "it trains them that results are unpredictable without an ORDER BY statement ... this kind of training is user-hostile". I disagree this is user-hostile because of what Arjun said: "it always works in testing on small data sizes, and only fails at scale". We *already* return data in unspecified order. That is, during testing, you can run a query and get back data in a different order each time based on how distsql performs. But this only happens if you are testing on a cluster with multiple nodes and with enough data that your SELECT covers multiple ranges.

If we were a single-node or single-master database whose characteristics were that we always returned data in the some order no matter what the size of the table or query was, then I would see the argument against this proposal. The current state is that we already do what this issue proposes, but only in some cases that are unpredictable to most users. And, even worse, may only occur in their production data because that's where they have an actual cluster of nodes with multiple ranges.

This proposal doesn't add new behavior, it makes the current behavior consistent regardless of how much data there is.

**Comment by benesch** [comment]
I have a semi-ancient working implementation of this here: https://github.com/cockroachdb/cockroach/compare/master...benesch:doshuffle

It significantly predates this discussion and so is only enabled if you `SET shuffle_unordered_results` in your shell. Obviously it's easy to instead turn it on all the time.

I tried to turn this on for logic tests and got completely and utterly distracted by how badly broken multi-line whitespace handling is in current logic tests. It's completely by accident that it works at all.

**Comment by jordanlewis** [comment]
We don't have plans to take this work on, so I'm going to close this.
